### PR TITLE
feat: Add support for cygwin/msys2/git-bash evironment

### LINKED ISSUE #590: Starship cannot be loaded from Fish inside Git-Bash
#### Current Behavior
When executing the Fish shell inside an instance of the Git Bash SDK in Windows, Starship cannot be loaded inside Fish because it doesn't follow the terminal's path syntax. 

`starship init fish` yields:
```fish
source ("C:\Users\Amalia\.cargo\bin\starship.exe" init fish --print-full-init | psub)⏎  
```
which will not work, because MSYS2 (the terminal host) expect Unix-style routes.

#### Expected Behavior
`starship init fish` should work in MSYS2.

#### Additional context/Screenshots
Since there are no Starship binaries for Windows, I installed Rust with `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh` inside the Git-Bash SDK window, and then issued `cargo install starship`.

#### Environment
- Starship version: 0.25.2
- Shell type: fish
- Shell version: 3.0.2
- Shell plugin manager: N/A
- Terminal emulator: Git-Bash SDK (https://github.com/git-for-windows/git-sdk-64), mintty 3.0.6 (x86_64-pc-msys)
- Operating system: Windows 10 1903

#### Relevant Shell Configuration
None, I have just installed the shell using `pacman`.

#### Starship Configuration
None, I have yet to initialize Starship.

#### Possible Solution
Use the plain executable name when constructing the source-able script.

### LINKED ISSUE #634: Support native windows (msys2) environment
#### Is your feature request related to a problem? Please describe.
Currently it's possible to build starship in native windows environment. Potentially this could be used in some environment like msys2 / cygwin, which allows shell env like zsh / bash.

Immediate problem I have noticed was path resolution to native binary when init shell could not work on windows path, which may need some interop.

#### Describe the solution you'd like

- Prebuilt binary for windows native
- Resolving path resolution & additional future fixes to support zsh / bash on msys2

### LINKED ISSUE #1354: Can't make Starship work on a Fish (Cygwin) on Windows environment
## Bug Report
Can't make Starship work on a Fish (Cygwin) on Windows environment

#### Current Behavior
Fish hangs when I start it after configuring `config.fish` to use Starship

gr82WXpH2Y

#### Expected Behavior
Should work

#### Additional context/Screenshots
After reading #1310 and #1341 , installing and verifying version with `/usr/local/bin/starship.exe --version`,
I now get another error:
```
fish: Unknown command: C:\\cygwin64\\usr\\local\\bin\\starship.exe
- (line 11):
    "C:\cygwin64\usr\local\bin\starship.exe" prompt --status=$exit_code --keymap=$keymap --cmd-duration=$starship_duration --jobs=(count (jobs -p))
    ^
in function 'fish_prompt'
in command substitution
```

I have contacted the fish team for help and following until [the end](https://github.com/fish-shell/fish-shell/issues/5151#issuecomment-646764438) of [my discussion here](https://github.com/fish-shell/fish-shell/issues/5151), it seems that there is a bug regarding paths.

> faho commented 33 minutes ago
> Okay, starship's windows version prints a broken path using windows-style \. I'm afraid that's a starship bug or at least incompatibility with cygwin.
>
> You'd have to check at what starship init fish --print-full-init prints, replace the broken paths and just put that directly in config.fish instead of going via source.

#### Possible Solution
Fix printing of paths so that the file is read correctly.

#### Environment
- Starship version: 0.42.0
- Shell type: fish
- Shell version: 3.1.0
- Shell plugin manager: None
- Terminal emulator: Cygwin (?) (C:\cygwin64\bin\mintty.exe)
- Operating system: Windows 10

#### Relevant Shell Configuration
\# ~/.config/fish/config.fish (please see why [here](https://github.com/fish-shell/fish-shell/issues/5151#issuecomment-646758518))
/usr/local/bin/starship.exe init fish --print-full-init | source

#### Starship Configuration
N/A

### PR DESCRIPTION
**Pull request by zoritle**
#### Description

see #1652 
Refactored the init logic.
Clean commit tree.
#### Motivation and Context

fixes #1354
fixes #634
fixes #590 

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [x] I have tested using **Windows**

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

**Comment by nesara-cmd** [comment]
0e826b5d21

File: src/init/mod.rs [hunk 1]
```diff
@@ -143,32 +189,31 @@ fi"#,
...
-        "powershell" => Some(PWSH_INIT),
-        "ion" => Some(ION_INIT),
+    let starship_path = StarshipPath::init()?;
+
+    match shell_name {
+        "bash" => print_script(BASH_INIT, &starship_path.sprint_unix()?),
+        "zsh" => print_script(ZSH_INIT, &starship_path.sprint_unix()?),
+        "fish" => print_script(FISH_INIT, &starship_path.sprint_unix()?),
+        "powershell" => print_script(PWSH_INIT, &starship_path.sprint()?),
+        "ion" => print_script(ION_INIT, &starship_path.sprint()?),
```
Comment (davidkna) [inline src/init/mod.rs:199]: I think `powershell` and `ion` should both use `sprint_unix`. PowerShell has a linux version and ion doesn't even run on windows.

**Comment by nesara-cmd** [comment]
https://github.com/starship/starship/commit/c5a543ec51ce705fc8add733592d027442defa1a

File: src/init/mod.rs (hunk 1 above)
Comment (zoritle) [reply src/init/mod.rs:199]: No. `sprint` directly return the platform native path string, which is the normal case.

`sprint_unix` return converted(on Windows)/native(on other platforms)  path string, is only used for some shells(bash/zsh/fish/etc) that runs on Windows through posix compatibility layer provided by cygwin.To be more precise, I renamed it to `sprint_posix`.

These shells can only accept posix path string (`/c/Windows/`) not native (`C:\Windows`), which cause the issues.

PowerShell is natively cross-platform and can handle native path on every platform. Ion though have not yet supported Windows, we shouldn't do additional thing except they decide to support Windows via cygwin runtime.

By the way, bash have been additionally patched to handle both Windows and posix path string, that's why people used git-bash with starship without problem, but zsh/fish not. To be consistent, we don't rely on this behavior and also apply conversion to it.

Comment (davidkna) [reply src/init/mod.rs:199]: Wouldn't PowerShell running via the Cygwin compatibly layer on windows still need the posix-style path? Also ion might not officially support windows, but I expect running it with Cygwin is possible.

Comment (zoritle) [reply src/init/mod.rs:199]: Cygwin is a compatibility layer specifically designed for C programs written against the posix api.
Powershell is built on .NET platform, which is intrinsic cross platform. I don't know if there're any way or any reason to build .NET runtime against cygwin...
Rust [did not support cygwin](https://users.rust-lang.org/t/does-rust-work-in-cygwin-if-so-how-can-i-get-it-working/25735) as a target, nor need to. Porting Ion to Windows is more possible than building it against cygwin since rust is a modern programming language.
Anyhow, since Ion currently unable to run on Windows, whether patch it or not makes no sense. I think keep it intact is more intuitive and avoid unexpected behavior in the future...

**Comment by nesara-cmd** [comment]
CMS100648-1)

**Comment by nesara-cmd** [comment]
CMS100648-1)

**Comment by davidkna** [comment]
I tried this in git bash and `starship.exe init bash` still returned the Windows path. Is this expected?

**Comment by zoritle** [comment]
> 
> 
> I tried this in git bash and `starship.exe init bash` still returned the Windows path. Is this expected? Do I need to use the GNU windows toolchain?

No, it should return converted path, and which rust toolchain you use does not matter.
For troubleshooting:
1. Ensure your are building from `feat/support_cygwin` branch, not master.
2. Ensure the `starship.exe` you called was built from source, not the official release installed by other tools (for example, scoop).
3.  Ensure you can call `cygpath.exe` inside git-bash, and `which cygpath` should return `/usr/bin/cygpath`.
4. Ensure git config `core.autocrlf` is not set to `true` before checkout, otherwise may cause line ending issues in embedded init script.

**Comment by davidkna** [comment]
Well seems to work now, must have done something wrong the last time.

File: src/init/mod.rs [hunk 2]
```diff
@@ -17,12 +17,55 @@ using whatever mechanism is available in the host shell--this two-phase solution
...
+            // Failed to execute cygpath.exe means there're not inside cygwin evironment,return directly.
+            Err(_) => {
+                return self.sprint();
+            }
+        };
+        let res = String::from_utf8(output.stdout);
+        let posix_path = match res {
+            Ok(ref cygpath_path) if output.status.success() => cygpath_path.trim(),
+            Ok(output) => {
+                eprintln!("Warning: getting unix path failure:{}, fallback.", output);
```
Comment (davidkna) [inline src/init/mod.rs:59]: Please use `log::warn!` (or better `log::error!`), add spaces and maybe reword the message to: `Failed to convert "{}" to unix path:\n{}`.

File: src/init/mod.rs [hunk 3]
```diff
@@ -17,12 +17,55 @@ using whatever mechanism is available in the host shell--this two-phase solution
...
+                str_path
+            }
+            Err(e) => {
+                eprintln!("Warning: invalid cygpath:{},fallback.", e);
```
Comment (davidkna) [inline src/init/mod.rs:63]: Same here

**Review by davidkna — CHANGES_REQUESTED** [review]

File: src/init/mod.rs [hunk 4]
```diff
@@ -17,12 +17,59 @@ using whatever mechanism is available in the host shell--this two-phase solution
...
+            }
+        };
+        let res = String::from_utf8(output.stdout);
+        let posix_path = match res {
+            Ok(ref cygpath_path) if output.status.success() => cygpath_path.trim(),
+            Ok(output) => {
+                log::warn!(
+                    "Failed to convert \"{}\" to unix path:\n{}",
+                    str_path,
+                    output
```
Comment (davidkna) [inline src/init/mod.rs:62]: cypaths error messages are on stderr.

Comment (davidkna) [inline src/init/mod.rs:62]:
```suggestion
            Ok(_) => {
                log::warn!(
                    "Failed to convert \"{}\" to unix path:\n{}",
                    str_path,
                    String::from_utf8_lossy(&output.stderr),
```

File: src/init/mod.rs [hunk 5]
```diff
@@ -17,12 +17,59 @@ using whatever mechanism is available in the host shell--this two-phase solution
...
+        let str_path = self.str_path()?;
+        let res = std::process::Command::new("cygpath.exe")
+            .arg(str_path)
+            .output();
+        let output = match res {
+            Ok(output) => output,
+            // Failed to execute cygpath.exe means there're not inside cygwin evironment,return directly.
+            Err(_) => {
+                return self.sprint();
+            }
```
Comment (davidkna) [inline src/init/mod.rs:53]: Might also make sense to combine with the match below.

```suggestion
            Err(e) if e.kind() == ErrorKind::NotFound => {
                return self.sprint();
            }
```

File: src/init/mod.rs [hunk 6]
```diff
@@ -17,12 +17,63 @@ using whatever mechanism is available in the host shell--this two-phase solution
...
+        if cfg!(not(target_os = "windows")) {
+            return self.sprint();
+        }
+        let str_path = self.str_path()?;
+        let res = std::process::Command::new("cygpath.exe")
+            .arg(str_path)
+            .output();
+        let output = match res {
+            Ok(output) => output,
+            // Failed to execute cygpath.exe means there're not inside cygwin evironment,return directly.
```
Comment (davidkna) [inline src/init/mod.rs:50]:
```suggestion
```

**Review by davidkna — APPROVED** [review]
LGTM, but please remove that duplicate comment.

Also please use regular commits for your progress. PRs are squashed when merging.

**Comment by andytom** [comment]
Thank you for you contribution @zoritle and thank you for reviewing @davidkna.
